ethcore/src/ethereum/ethash.rs

```diff
@@ -186,24 +205,38 @@ impl Engine<EthereumMachine> for Arc<Ethash> {
 		let author = *LiveBlock::header(&*block).author();
 		let number = LiveBlock::header(&*block).number();
 
+		// Applies EIP-649 reward.
 		let reward = if number >= self.ethash_params.eip649_transition {
 			self.ethash_params.eip649_reward.unwrap_or(self.ethash_params.block_reward)
 		} else {
 			self.ethash_params.block_reward
 		};
 
+		// Applies ECIP-1017 eras.
 		let eras_rounds = self.ethash_params.ecip1017_era_rounds;
 		let (eras, reward) = ecip1017_eras_block_reward(eras_rounds, reward, number);
 
 		let n_uncles = LiveBlock::uncles(&*block).len();
 
-		// Bestow block reward
+		// Bestow block rewards.
 		let result_block_reward = reward + reward.shr(5) * U256::from(n_uncles);
 		let mut uncle_rewards = Vec::with_capacity(n_uncles);
 
-		self.machine.add_balance(block, &author, &result_block_reward)?;
+		if number >= self.ethash_params.mcip3_transition {
+			let miner_reward = self.ethash_params.mcip3_miner_reward.unwrap();
+			let ubi_contract = self.ethash_params.mcip3_ubi_contract;
+			let ubi_reward = self.ethash_params.mcip3_ubi_reward.unwrap();
+			let dev_contract = self.ethash_params.mcip3_dev_contract;
+			let dev_reward = self.ethash_params.mcip3_dev_reward.unwrap();
```

Comment:
we don't use `unwrap()`, favor `expect()` with a proof that it will never panic. That these fields are either all set or all not should be checked at spec parsing time. Grouping them into a `struct mcip3` parsed optionally would make that logic a lot easier.

Reply:
This makes sense. I have rewritten the logic, now a malformed mcip3 spec will just behave like a default block reward scheme.